CRM/Contact/BAO/GroupContactCache.php

```diff
@@ -618,19 +604,52 @@ public static function load(&$group, $force = FALSE) {
       $childrenIDs = explode(',', $group->children);
       foreach ($childrenIDs as $childID) {
         $contactIDs = CRM_Contact_BAO_Group::getMember($childID, FALSE);
-        //Unset each contact that is removed from the parent group
+        // Unset each contact that is removed from the parent group
         foreach ($removed_contacts as $removed_contact) {
           unset($contactIDs[$removed_contact]);
         }
+        if (empty($contactIDs)) {
+          // This child group has no contact IDs so we don't need to add them to
+          continue;
+        }
         $values = [];
         foreach ($contactIDs as $contactID => $dontCare) {
           $values[] = "({$groupID},{$contactID})";
         }
-
-        self::store([$groupID], $values);
+        $str = implode(',', $values);
+        CRM_Core_DAO::executeQuery("INSERT IGNORE INTO $tempTable (group_id, contact_id) VALUES $str");
       }
     }
 
+    // grab a lock so other processes don't compete and do the same query
+    $lock = Civi::lockManager()->acquire("data.core.group.{$groupID}");
+    if (!$lock->isAcquired()) {
+      // this can cause inconsistent results since we don't know if the other process
+      // will fill up the cache before our calling routine needs it.
+      // however this routine does not return the status either, so basically
+      // its a "lets return and hope for the best"
+      return;
+    }
+
+    // Don't call clearGroupContactCache as we don't want to clear the cache dates
+    // The will get updated by updateCacheTime() below and not clearing the dates reduces
+    // the chance that loadAll() will try and rebuild at the same time.
+    $clearCacheQuery = "
+    DELETE  g
+      FROM  civicrm_group_contact_cache g
+      WHERE  g.group_id = %1 ";
+    $params = [
+      1 => [$groupID, 'Integer'],
+    ];
+    CRM_Core_DAO::executeQuery($clearCacheQuery, $params);
+
+    CRM_Core_DAO::executeQuery(
+      "INSERT IGNORE INTO civicrm_group_contact_cache (contact_id, group_id)
```

Comment:
We shouldn't need INSERT IGNORE here

Reply:
No we shouldn't, but we do! Without INSERT IGNORE I was still seeing occasional database exceptions because of duplicate entries (every few hours or so). It suggests that there is still a hidden contention issue somewhere deep down but I couldn't find out where.  It would be good to add a comment to that effect here to be clear why we have INSERT IGNORE but solving that problem shouldn't hold up this PR and could be a follow-up once we've sorted out the major deadlocks via this PR.  It is worth noting that INSERT IGNORE is significantly slower than INSERT by itself so would be nice to resolve at some point!

Reply:
Actually just looking - this is a copy of the code in self::clearGroupContactCache($groupID); - which is called a few lines earlier - so this isn't a behaviour change - just a duplication